tmt/recipe.py

```diff
@@ -0,0 +1,186 @@
+from typing import TYPE_CHECKING, Any, Optional, cast
+
+import tmt.utils
+from tmt.container import SerializableContainer, container, field
+from tmt.log import Logger
+from tmt.steps import Step, StepData
+from tmt.steps.discover import Discover
+from tmt.steps.execute import Execute
+from tmt.utils import Common, Environment, FmfContext, Path
+
+if TYPE_CHECKING:
+    import tmt.base
+    from tmt.base import Plan, Run, Test, _RawAdjustRule
+
+
+def _serialize_tests(tests: list['Test']) -> list[dict[str, Any]]:
+    serialized_tests: list[dict[str, Any]] = []
+    for test in tests:
+        serialized_test = test._export(include_internal=True)
+        # Replace the modified environment with the original one
+        serialized_test['environment'] = test._original_fmf_environment.to_fmf_spec()
+        serialized_tests.append(serialized_test)
+    return serialized_tests
+
+
+@container
+class _Step(SerializableContainer):
+    enabled: bool
+    phases: list[StepData] = field(
+        serialize=lambda value: [phase.to_serialized() for phase in value],
+    )
+
+    @classmethod
+    def from_plan(cls, plan: 'Plan', step_name: str) -> '_Step':
+        step = getattr(plan, step_name)
+        if isinstance(step, Discover):
+            return _DiscoverStep(
+                enabled=bool(step.enabled),
+                phases=cast(Step, step).data if step.enabled else [],
+                tests=[test_origin.test for test_origin in step.tests()],
+            )
+        if isinstance(step, Execute):
+            return _ExecuteStep(
+                enabled=bool(step.enabled),
+                phases=cast(Step, step).data if step.enabled else [],
+                results_path=(step.step_workdir / 'results.yaml').relative_to(step.run_workdir),
+            )
+        if isinstance(step, Step):
+            return _Step(enabled=bool(step.enabled), phases=step.data if step.enabled else [])
+
+        raise ValueError(f"Step '{step_name}' is not defined in plan '{plan.name}'")
+
+
+@container
+class _DiscoverStep(_Step):
+    tests: list['Test'] = field(serialize=_serialize_tests)
```

Comment:
This should probably be the recipe-ized form of `Test` instances: you have `_RecipeRun`, and `_RecipePlan`, and `_Step` (which, BTW, should probably be called `_RecipeStep` to follow the suit), and they are all these "snapshots" of live run, step and plan objects. But `_Step.tests` isn't. That's kind of exceptional, I'd recommend filling it with what `_serialize_tests()` produces, so that all levels of the hierarchy are already recipe-ized.

Either that, or reimplement `_Run`, `_Plan` and `_Step` to recipe-ize themselves when being serialized, transparently, which is exactly what's happening to `_DiscoverStep.tests`. That would also be possible, but a bigger change at this moment.